content/blog/c-off-ramp.md

```diff
@@ -0,0 +1,295 @@
+
+I was also able to offer my old-school expertise to the Vala project itself: I converted the test suite to use Automake’s relatively recent ability to run tests in parallel.
+
+#### Falling between two stools
+
+Vala itself mostly insulates you from the memory-management woes that bedevil C, but mid-translation there is still C code, so you have to be careful if you want to keep a working program. It’s always tempting to just plough on with the translation and hope that crashes will disappear, but I try not to give in, as that can bury bugs deeper and make them harder to find later. So, expect to spend some time studying `valac`-generated C, and revising your knowledge of `owned` and other esoteric Vala features.
+
+For me, this is particularly relevant to Enchant, as a C library whose API and ABI I want to maintain. I am happy to report that I have been mostly successful, and at the time of writing I am tracking down some problems with object lifetime management, but otherwise the Vala implementation seems to perfectly replicate the C.
+
+Another area of awkwardness is strings: as I hinted above, C often uses pointers to refer to string positions, whereas in Vala it’s usual to use integer offsets. This code is fairly easy to convert once you get the hang of it. Since C *can* use offsets instead of pointers, you can always rewrite the code in two stages, first changing the C to use offsets, and then translating it into Vala.
+
+Here’s an example of the translation from pointers in C to offsets in Vala. The C version refers to positions in strings with pointers:
+
+```c
+char * enchant_iso_639_from_tag (const char * const dict_tag) {
+    char * new_tag = strdup (dict_tag);
+    if (new_tag == NULL)
+        return NULL;
+    char * needle = strchr (new_tag, '_');
+
+    if (needle != NULL)
+        *needle = '\0';
+
+    return new_tag;
+}
+```
+
+In Vala, we use string offsets:
+
+```vala
+static string iso_639_from_tag(string dict_tag) {
+    return dict_tag.substring(0, dict_tag.index_of_char('_'));
+}
+```
+
+The comparison here is perhaps obscured by the simplification obtained in Vala from automatic memory management, which is nothing to do with string handling *per se*.
+
+#### Binding challenges
+
+I had a number of particular challenges. Some of them may be relevant to other developers, but I detail them here mostly to illustrate that Vala can bind just about anything if you try!
+
+##### `config.h`
+
+Like any Autoconf-based build system, mine produce a `config.h` header of various `#define`d symbols, some of which my C code uses. These I simply replicate in a `Config` module, following a pattern by Jürg Billeter, one of the original authors of Vala:
```

Comment:
actually the example isn't all that helpful. I think best would be to just put a link here: https://gitlab.gnome.org/GNOME/vala/-/blob/main/vapi/config.vapi?ref_type=heads

and also I would remove the part about a "pattern" being used/from whom it originates.

Reply:
I wanted to give credit to Jürg, but are you saying that it's not a significant idea, so doesn't need credit? Good idea about the link, I'll use that.